src/wrappedGraphs/graphviews.jl

```diff
@@ -0,0 +1,125 @@
+"""
+    ReverseView{T<:Integer,G<:AbstractGraph} <: AbstractGraph{T}
+
+A graph view that wraps a directed graph and reverse the direction of every edge.
+
+# Examples
+```jldoctest
+julia> using Graphs
+
+julia> g = SimpleDiGraph(2);
+
+julia> add_edge!(g, 1, 2);
+
+julia> rg = ReverseView(g);
+
+julia> neighbors(rg, 1)
+Int64[]
+
+julia> neighbors(rg, 2)
+1-element Vector{Int64}:
+ 1
+```
+"""
+struct ReverseView{T<:Integer,G<:AbstractGraph} <: AbstractGraph{T}
+    g::G
+
+    @traitfn ReverseView{T,G}(g::::(IsDirected)) where {T<:Integer,G<:AbstractGraph{T}} =
+        new(g)
+    @traitfn ReverseView{T,G}(g::::(!IsDirected)) where {T<:Integer,G<:AbstractGraph{T}} =
+        throw(ArgumentError("Your graph needs to be directed"))
+end
+
+ReverseView(g::G) where {T<:Integer,G<:AbstractGraph{T}} = ReverseView{T,G}(g)
+
+wrapped_graph(g::ReverseView) = g.g
+
+Graphs.is_directed(::ReverseView{T,G}) where {T,G} = true
+Graphs.is_directed(::Type{<:ReverseView{T,G}}) where {T,G} = true
+
+Graphs.edgetype(g::ReverseView) = Graphs.edgetype(g.g)
+Graphs.has_vertex(g::ReverseView, v) = Graphs.has_vertex(g.g, v)
+Graphs.ne(g::ReverseView) = Graphs.ne(g.g)
+Graphs.nv(g::ReverseView) = Graphs.nv(g.g)
+Graphs.vertices(g::ReverseView) = Graphs.vertices(g.g)
+Graphs.edges(g::ReverseView) = (reverse(e) for e in Graphs.edges(g.g))
+Graphs.has_edge(g::ReverseView, s, d) = Graphs.has_edge(g.g, d, s)
+Graphs.inneighbors(g::ReverseView, v) = Graphs.outneighbors(g.g, v)
+Graphs.outneighbors(g::ReverseView, v) = Graphs.inneighbors(g.g, v)
+
+"""
+    UndirectedView{T<:Integer,G<:AbstractGraph} <: AbstractGraph{T}
+
+A graph view that wraps a directed graph and consider every edge as undirected.
+
+# Examples
+```jldoctest
+julia> using Graphs
+
+julia> g = SimpleDiGraph(2);
+
+julia> add_edge!(g, 1, 2);
+
+julia> ug = UndirectedView(g);
+
+julia> neighbors(ug, 1)
+1-element Vector{Int64}:
+ 2
+
+julia> neighbors(ug, 2)
+1-element Vector{Int64}:
+ 1
+```
+"""
+struct UndirectedView{T<:Integer,G<:AbstractGraph} <: AbstractGraph{T}
+    g::G
+    ne::Int
```

Comment:
Precalculating `ne` is a bit dangerous as one might change the underlying graph and then the view would be invalid. If we cache this value then we should be very clear about that in the documentation.

I think it might be worth not to cache it and then make sure that all our algorithms calculate `ne` only once.